stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -0,0 +1,1385 @@
+        let miner_auth = coinbase_tx.get_origin();
+        let miner_addr = miner_auth.get_address(mainnet);
+
+        let recipient = if epoch_id >= StacksEpochId::Epoch21 {
+            // pay to tx-designated recipient, or if there is none, pay to the origin
+            match coinbase_tx.try_as_coinbase() {
+                Some((_, recipient_opt)) => recipient_opt
+                    .cloned()
+                    .unwrap_or(miner_addr.to_account_principal()),
+                None => miner_addr.to_account_principal(),
+            }
+        } else {
+            // pre-2.1, always pay to the origin
+            miner_addr.to_account_principal()
+        };
+
+        // N.B. a `MinerPaymentSchedule` that pays to a contract can never be created before 2.1,
+        // per the above check (and moreover, a Stacks block with a pay-to-alt-recipient coinbase would
+        // not become valid until after 2.1 activates).
+        let miner_reward = MinerPaymentSchedule {
+            address: miner_addr,
+            recipient,
+            block_hash: block_hash.clone(),
+            consensus_hash: block_consensus_hash.clone(),
+            parent_block_hash: parent_block_hash.clone(),
+            parent_consensus_hash: parent_consensus_hash.clone(),
+            coinbase: coinbase_reward_ustx,
+            tx_fees: MinerPaymentTxFees::Nakamoto { parent_fees },
+            burnchain_commit_burn,
+            burnchain_sortition_burn,
+            miner: true,
+            stacks_block_height: block_height,
+            vtxindex: 0,
+        };
+
+        Ok(miner_reward)
+    }
+
+    /// Return the total ExecutionCost consumed during the tenure up to and including
+    ///  `block`
+    pub fn get_total_tenure_cost_at(
+        conn: &Connection,
+        block: &StacksBlockId,
+    ) -> Result<Option<ExecutionCost>, ChainstateError> {
+        let qry = "SELECT total_tenure_cost FROM nakamoto_block_headers WHERE index_block_hash = ?";
+        conn.query_row(qry, &[block], |row| row.get(0))
+            .optional()
+            .map_err(ChainstateError::from)
+    }
+
+    /// Return the total transactions fees during the tenure up to and including
+    ///  `block`
+    pub fn get_total_tenure_tx_fees_at(
+        conn: &Connection,
+        block: &StacksBlockId,
+    ) -> Result<Option<u128>, ChainstateError> {
+        let qry = "SELECT tenure_tx_fees FROM nakamoto_block_headers WHERE index_block_hash = ?";
+        let tx_fees_str: Option<String> =
+            conn.query_row(qry, &[block], |row| row.get(0)).optional()?;
+        tx_fees_str
+            .map(|x| x.parse())
+            .transpose()
+            .map_err(|_| ChainstateError::DBError(DBError::ParseError))
+    }
+
+    /// Return a Nakamoto StacksHeaderInfo at a given tenure height in the fork identified by `tip_index_hash`
+    pub fn get_header_by_tenure_height(
+        tx: &mut StacksDBTx,
+        tip_index_hash: &StacksBlockId,
+        tenure_height: u64,
```

Comment:
Might be worth a comment to say that `tenure_height` is equivalent to the Stacks block height pre-Nakamoto.

Reply:
Might also be worth a comment to explain that in Nakamoto, this gets the `StacksHeaderInfo` for the first block in a Nakamoto tenure.